### ISSUE TITLE
Suggested API change – extend/collect should panic when capacity overflows

### ISSUE THREAD

**Issue by whatisaphone** [bluss/arrayvec#163]
I noticed an API inconsistency. `ArrayVec::push` panics if you overflow the capacity, but `ArrayVec::extend` silently drops elements if you overflow the capacity.

I prefer the panicking behavior, since otherwise you run the risk of accidentally losing iterator elements. imo that's not an intuitive failure mode. I'd suggest changing `extend` to panic on overflow, and introduce a fallible `try_extend` (after all, you already have `try_extend_from_slice`), `try_from_iterator`, and probably `try_collect` and a trait to go with it.

This is a breaking change. The migration path would be pretty straightforward:

```rust
// previous version
arrayvec.extend(iterator);

// next version
arrayvec.extend(iterator.take(arrayvec.remaining_capacity());
// -- OR --
let _ignore_result = arrayvec.try_extend(iterator);
```

**Comment by bluss** [comment]
I'm inclined to agree, but would like to hear other's thoughts as well. It's not great that extend/fromiterator may panic, but there are only these two (bad) options.

**Comment by tbu-** [comment]
I'm also in favor of panicking.

**Comment by bluss** [comment]
I think this change could be done for the next breaking version update - looking like the const generics update

**Comment by bluss** [comment]
The behaviour change is implemented in #175, but .try_extend() is not implemented yet (that's a new feature, not a behaviour change). A new issue is needed for that.

I'm a bit unsure about the `try_extend` name because I'm not so keen to have a "try" method that can return an error, but still have an effect in the error case. Since it's an iterator we wouldn't know its length until we iterate it. So maybe it could be called something else than "try_extend". For example, "try_extend_from_slice" doesn't partly extend in the error case - the arrayvec is unchanged on error.
